docs/structured-streaming-programming-guide.md

```diff
@@ -0,0 +1,888 @@
+the final `wordCounts` DataFrame is the result table. Note that the query on 
+streaming `lines` DataFrame to generate `wordCounts` is *exactly the same* as 
+it would be a static DataFrame. However, when this query is started, Spark 
+will continuously check for new data from the socket connection. If there is 
+new data, Spark will run an “incremental” query that combines the previous 
+running counts with the new data to compute updated counts, as shown below.
+
+![Model](img/structured-streaming-example-model.png)
+
+This model is significantly different from many other stream processing 
+engines. Many streaming system require the user to maintain running 
+aggregations themselves, thus having the reason about fault-tolerance, and 
+data consistency (at-least-once, or at-most-once, or exactly-once). In this 
+model, Spark is responsible for updating the Result Table when there is new 
+data, thus relieving the users from reasoning about it. As an example, let’s 
+see how this model handles event-time based processing and late arriving data.
+
+## Handling Event-time and Late Data
+Event-time is the time embedded in the data itself. For many applications, you may want to do operate using this event-time. For example, if you want to get the number of events generated by IoT devices every minute, then you probably want to use the time when the data was generated (that is, event-time in the data), rather than the time Spark receives them. This event-time is very naturally expressed in this model -- each event from the devices is a row in the table, and event-time a column value in the row. This allows window-based aggregations (e.g. number of event every minute) to be just a special type of grouping and aggregation on the even-time column -- each time window is a group and each row can belong to multiple windows/groups. Therefore, such event-time-window-based aggregation queries can be defined consistently on both a static dataset (e.g. from collected device events logs) as well as on streaming dataset, making the life of the user much easier.
+
+Furthermore this model naturally handles data that has arrived later than expected based on its event-time. Since Spark is updating the Result Table, it has full control over updating/cleaning up the aggregates when there is late data. While not yet implemented in Spark 2.0 yet, event-time watermarking will be used to manage this data. These are explained in more details in the Window Operations section.
+
+## Fault Tolerance Semantics
+Delivering end-to-end exactly-once semantics was one of key goals behind the design of Structured Streaming. To achieve that, we have designed the Structured Streaming sources, the sinks and the execution engine to reliably track the exact progress of the processing so that it can handle any kind of failure by restarting and/or reprocessing. Every streaming source is assumed to have offsets (similar to Kafka offsets, or Kinesis sequence numbers)
```

Comment:
If you aren't going to qualify exactly-once in the intro, need to do it here.  Producing into Kafka isn't exactly-once without additional work, for instance.